Dev/controls

nadimasmar opened the pull request:
This PR adds controls and state estimation packages. Both packages have been tested in the pool and provide the following functionality:

1. IMU data transformation into the AUV frame. Extensively tested in the pool
2. DVL data transformation into the AUV frame. Modereately tested in the pool.
3. Depth Sensor data transformation into the AUV frame. Extensively tested in the pool
4. Depth Control. Dive, hover, and surface capability that is accurate and precise. Extensively tested in the pool. 
5.  Attitude Control. Yaw accurately and precisely to any possible heading. Corrective Roll and Pitch in case of a disturbance. Roll and Pitch maneuvers (such as barrel rolls) are not supported, only corrections to the neutral state in case of a disturbance.  Extensively tested in pool
6. Planar ( X & Y) control. Translate along the X and Y axes accurately and precisely. Lightly tested in the pool. Only straight maneuvers have been tested, diagonal movements have not been tested.

nadimasmar committed:
Initial Commit

nadimasmar committed:
Update README.md

nadimasmar committed:
Debugged setup, commented out recording, properly imported thrust_mapper_utils

Stuart-Klenner committed:
inital mechtest2.py file, variables need to be updated though but wanted to check if overall format is correct

Stuart-Klenner committed:
Updated commenting in file

Stuart-Klenner committed:
added mechtest2 launch file and small comment changes to mechtest2 file

hutchallison committed:
added drytest.py and drytest.launch.py files converted to ROS2, added line to setup.py to account for new file

hutchallison committed:
added drytest.launch.py file

nadimasmar committed:
added mechtest2 and drytest launch files to setup.py

nadimasmar committed:
Merge remote-tracking branch 'origin/custom_messages' into thrust_allocation
Adding messages in order to colcon buold propulsion package successfully.

nadimasmar committed:
fixed publisher arguments, parameters declaration

hutchallison committed:
Debugged drytest.py

nadimasmar committed:
Switched PWM messages to standard Int16MultiArray

nadimasmar committed:
Merge branch 'main' into dev/thrust_allocation

hutchallison committed:
Finished debugging drytest.py

nadimasmar committed:
added micro-ros-agent to propulsion.launch.py

Stuart-Klenner committed:
added mechtest2 launch code

hutchallison committed:
changed the /propulsion/microseconds message type in drytest.py to standard messasge Int16MultiArray

hutchallison committed:
deleted "drytest.launch.py" as drytest can not be run from the launch file due to the use of the input function, use ros2 run propulsion drytest command to run instead

Stuart-Klenner committed:
replaced custom messages in mechtest2 file with standard messages

vbigattini committed:
Updated allocation matrix and launch params dx and dy for off-centre CG in xy plane

vbigattini committed:
Fixed parameter declaration

hutchallison committed:
Added stopping command to keyboard control in drytest.py, deleted drytest.launch.py dependancy in setup.py to fix build issue

hutchallison committed:
added some user input error catching sections to drytest.py

thirteenlizards committed:
Merge branch 'main' into dev/thrust_allocation

nadimasmar committed:
Merge branch 'main' into dev/thrust_allocation

JPGC04 committed:
Normalize line endings to LF

Stuart-Klenner committed:
Update setup.py

nadimasmar committed:
Merge branch 'main' into dev/thrust_allocation

nadimasmar committed:
Replace Thruster Allocation Matrix image

Updated the image link for the Thruster Allocation Matrix in the README.

nadimasmar committed:
Removed redundant mounting direction array.

nadimasmar committed:
Merge branch 'dev/thrust_allocation' of https://github.com/mcgill-robotics/AUV-2026 into dev/thrust_allocation

hutchallison committed:
added option five to select a thruster to test and a duration to test it for to help speed up pool tests

Stuart-Klenner committed:
Create imuprocessing

nadimasmar committed:
Write README for sensor data processing details

Expanded README to include detailed sensor data processing information, including IMU, depth sensor, and DVL processing. Added sections for usage, nodes, published and subscribed topics, installation instructions, and dependencies.

nadimasmar committed:
Fixed thruster arming messages.

nadimasmar committed:
Clarify parameter descriptions in README.md

Updated parameter descriptions for clarity.

Stuart-Klenner committed:
created imuprocessing.h file

created imuprocessing.h file. Takes acceleration readings from imu sensor and derives acceleration in the pool frame of reference. Requires nadim to add helper function to create rotation matrix from quarternion input from imu.

nadimasmar committed:
Merge branch 'main' into dev/thrust_allocation

vbigattini committed:
Implementing controls branch

Stuart-Klenner committed:
imu processing pipeline

vbigattini committed:
Created basic logic of PID controller in z-direction

nadimasmar committed:
docs: Update README with detailed IMU and depth sensor processing explanations

nadimasmar committed:
Fix formatting of mathematical equation in README

nadimasmar committed:
Imu processor, package restructuring

nadimasmar committed:
Added sensor node, and finished logic for data processors.

nadimasmar committed:
Merge remote-tracking branch 'origin/main' into dev/llmweapons

nadimasmar committed:
Enhance README with sensor_node details

Updated README.md to provide details about the sensor_node, including inputs, outputs, published and subscribed topics, dependencies, building, and running instructions.

nadimasmar committed:
fix: Update CMakeLists.txt to use Eigen3_cmake_module and make sensor_node run regardless of sim.

nadimasmar committed:
Implement control nodes and launch files, including attitude and depth controllers, and superimposer

nadimasmar committed:
Refactor launch files and superimposer parameters.

nadimasmar committed:
Clarify launch command for sensor data processors

Updated the README to clarify the launch command for sensors.

nadimasmar committed:
Fix: rotate depth effort in superimposer. Add README

nadimasmar committed:
Added control law documentation.

nadimasmar committed:
Corrected topic name, small README update

nadimasmar committed:
spelling fix

nadimasmar committed:
refactor: Simplify IMU processing by using precomputed quaternion inverses

nadimasmar committed:
Updated comments for clarity

WilliamZhang2205 committed:
Corrected sensors launch file by making it more clean and organized with the new nodes. JP approved commit message uwu

nadimasmar committed:
Dev/thrust allocation (#25)

Co-authored-by: Stuart-Klenner <stuartjklenner@gmail.com>
Co-authored-by: hutchallison <aahutchings06@gmail.com>
Co-authored-by: Vincenzo Bigattini <vbigattini2004@gmail.com>
Co-authored-by: anna joy <146257353+thirteenlizards@users.noreply.github.com>
Co-authored-by: John-Paul Chouery <johnpaul.chouery@gmail.com>

nadimasmar committed:
Dev/thrust allocation (#25)

Co-authored-by: Stuart-Klenner <stuartjklenner@gmail.com>
Co-authored-by: hutchallison <aahutchings06@gmail.com>
Co-authored-by: Vincenzo Bigattini <vbigattini2004@gmail.com>
Co-authored-by: anna joy <146257353+thirteenlizards@users.noreply.github.com>
Co-authored-by: John-Paul Chouery <johnpaul.chouery@gmail.com>

nadimasmar committed:
Refactored package: Each processor is now a `Node`  + new state aggregator

nadimasmar committed:
Change parameter type

nadimasmar committed:
Merge remote-tracking branch 'origin/main' into dev/llmweapons

nadimasmar committed:
Merge branch 'main' into dev/controls

nadimasmar committed:
change README

Stuart-Klenner committed:
Adding DVL data processing documentation

Stuart-Klenner committed:
Changed  DVL data processing documentation

Stuart-Klenner committed:
Changed  DVL data processing documentation

Stuart-Klenner committed:
Changed  DVL data processing documentation

nadimasmar committed:
Added link to external and tweaked notation  to follow Textbook

Updated the DVL processing section to clarify the transformation matrix notation and added a link to relevant documentation.

nadimasmar committed:
Corrected free acceleration calculation. Adopted +w convention in quaternions

nadimasmar committed:
Merge branch 'dev/llmweapons' of https://github.com/mcgill-robotics/AUV-2026 into dev/llmweapons

nadimasmar committed:
Corrected naming convention and IMU raw data frame of reference

nadimasmar committed:
Small fix: Class constructor

nadimasmar committed:
Merge branch 'dev/llmweapons' into dev/controls

nadimasmar committed:
refactored sensors package to follow better coding practices

nadimasmar committed:
refactored sensors package to follow better coding practices

ParseDotEXE commented on ros2_ws/src/controls/controls/pid.py:
```diff
@@ -0,0 +1,34 @@
+
+
+                self.integral_error += self.position_error * time_step
+                self.integral_error = np.clip(self.integral_error, -self.I_MAX, self.I_MAX)
```
if I_MAX is set to 0.0 this gives -> np.clip(self.intergral_error, -0.0, 0.0)
This caps the Integral error to exactly 0 -> could be handled differently maybe

ParseDotEXE commented on ros2_ws/src/controls/include/controls/attitude_controller.hpp:
```diff
@@ -0,0 +1,76 @@
+                Vec3 w_vi_;
+
+                //Target state variables
+                quatd q_vi2_; 
```
General comments for minor efficiency improvements:
1- Adding a quaternion validation check
2- This one is probably not that necessary but adding timeout handling for when the IMU data comes in choppy

ParseDotEXE commented on ros2_ws/src/controls/include/controls/superimposer.hpp:
```diff
@@ -0,0 +1,73 @@
+    using Mat3 = Eigen::Matrix3d;
+    using Vec3 = Eigen::Vector3d;
+
+    class superimposer: public rclcpp::Node
```
follow class naming conventions
class name unclear of what the class does

ParseDotEXE commented on ros2_ws/src/controls/launch/attitude_controller.launch.py:
```diff
@@ -0,0 +1,29 @@
+# launch/thrust_mapper.launch.py
```
does this launch attitude controller or thrust mapper?

ParseDotEXE commented on ros2_ws/src/controls/scripts/depth_controller.py:
```diff
@@ -0,0 +1,86 @@
+
+    def control_loop_callback(self):
+        # Compute PID errors
+        self.pid.compute_errors(self.setpoint_depth, self.current_depth, self.previous_depth, self.time_step)
```
This and the next line could be merged into one

Stuart-Klenner committed:
dvl processing draft files

nadimasmar committed:
Corrected attitude controller quaternion convention, put initial estimate for net buoyancy

nadimasmar committed:
initial buoyancy estimate

nadimasmar committed:
README update and package.xml fix

nadimasmar committed:
Corrected treatment of depth as +Z position. Depth is -Z position

nadimasmar committed:
Merge branch 'dev/llmweapons' into dev/controls

Stuart-Klenner committed:
Updated dvl processor hpp and cpp files and cmakelists file

Stuart-Klenner committed:
Updated dvl cpp, hpp, and cmakelists files

Stuart-Klenner committed:
Updated formatting and headers

nadimasmar committed:
Adjusted Topic names for output effort

nadimasmar committed:
Normalize all line endings to LF

nadimasmar committed:
Added .gitattributes to make all line endings LF.

nadimasmar committed:
Commented out line in build.sh

nadimasmar committed:
Switched Hard-coded paths with dynamic ones.
Co-authored-by: Sohaib Kaidali  <sohaib.kaidali@mail.mcgill.ca>

nadimasmar committed:
Merge branch 'dev/llmweapons' into dev/controls
Keeping controls branch up-to-date with sensor package updates

nadimasmar committed:
Refactor launch file installation paths and update quaternion variable names for clarity. Added bringup folder

nadimasmar committed:
temporary debugging changes

ParseDotEXE commented on ros2_ws/src/controls/src/attitude_controller.cpp:
```diff
@@ -0,0 +1,137 @@
+        this->declare_parameter<double>("P_wx", 1.0);
+        this->declare_parameter<double>("P_wy", 1.0);
+        this->declare_parameter<double>("P_wz", 1.0);
+        this->declare_parameter<double>("buoyancy", 278.0); // Newtons
```
good unit declaration

hutchallison commented on build.sh:
```diff
@@ -30,7 +30,7 @@ if [ -d "ros2_ws/src" ]; then
   fi
   cd ros2_ws
-  rosdep install --from-paths src --ignore-src -r -y --skip-keys="zed zed_msgs zed_components"
+  # rosdep install --from-paths src --ignore-src -r -y --skip-keys="zed zed_msgs zed_components"
```
Can we add a comment for why this was removed, "overwrites other necessary dependencies"

JPGC04 replied:
was it requested by vision people to remove this line?

ParseDotEXE commented on ros2_ws/src/controls/src/attitude_controller.cpp:
```diff
@@ -0,0 +1,137 @@
+        this->get_parameter("r_bv", r_bv_);
+        this->get_parameter("control_loop_hz", control_loop_hz_);
+
+        q_iv_ = quatd::Identity(); // Initial orientation: identity quaternion
```
Uninitialized control before first IMU message - Control loop starts immediately but q_iv_ and w_iv_ only update on first IMU callback. Should either wait for first message or add validity check.

fix: wait for the first message

ParseDotEXE commented on ros2_ws/src/controls/src/attitude_controller.cpp:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    void AttitudeController::imu_callback(const imu_msg::SharedPtr msg)
+    {
```
// In imu_callback
eg : imu_received_ = true;

ParseDotEXE commented on ros2_ws/src/controls/src/attitude_controller.cpp:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    void AttitudeController::control_loop_callback()
+    {
```
// In control_loop_callback
eg: if (!imu_received_) return; // Skip until IMU available

ParseDotEXE commented on ros2_ws/src/controls/src/superimposer.cpp:
```diff
@@ -0,0 +1,124 @@
+#include "controls/superimposer.hpp"
+
```
include a documentation to provide what the class does, what the methods do and what are the params

ParseDotEXE commented on ros2_ws/src/controls/src/superimposer.cpp:
```diff
@@ -0,0 +1,124 @@
+           msg->orientation.w,
+           msg->orientation.x,
+           msg->orientation.y,
+           msg->orientation.z
```
missing normalization

ParseDotEXE commented on ros2_ws/src/controls/src/superimposer.cpp:
```diff
@@ -0,0 +1,124 @@
+       );
+
+       // Initialize publisher
+       pub_effort_ = this->create_publisher<wrench_msg>("/controls/total_effort", 1);
```
README says output should be /controls/combined_effort?

hutchallison commented on ros2_ws/src/sensors/params/sensors_frames.yaml:
Nicely Commented

nadimasmar committed:
Launch directory properly placed under install/controls/share/controls/launch and add exec permissions to depth_controller.py

JPGC04 commented on .vscode/c_cpp_properties.json:
Do not commit this file please

JPGC04 commented on ros2_ws/src/bringup/launch/bringup.launch.py:
Do we want a separate sim.launch file similarly to in auv-2025? This is not relevant to this PR specifically, but it is a point to discuss for when more sim stuff gets merged.

JPGC04 commented on ros2_ws/src/controls/launch/depth_controller.launch.py:
```diff
@@ -0,0 +1,23 @@
+                executable='depth_controller.py',
+                name='depth_controller',
+                output='screen',
+                parameters=[{
```
I don't know if I missed another file with those parameters, but would be nice to have KP KD and KI params for the sim cause the tuning will probably be a bit different in sim. Like you could have a sim condition for pids.

JPGC04 commented on ros2_ws/src/propulsion/package.xml:
```diff
@@ -12,8 +12,6 @@
   <url type="repository">https://github.com/mcgill-robotics/AUV-2026</url>
   <author email="nadim.asmar@mail.mcgill.ca">Nadim Asmar</author>
 
-<buildtool_depend>ament_python</buildtool_depend>
```
Is this package not required?

JPGC04 commented on ros2_ws/src/sensors/launch/sensors.launch.py:
The physical sensor launches in my opinion should be on their own, for example the display microros agent should be launched in the depth process launchfile. same for imu and eventually dvl.  The problem with this file is that if the display fails to launch, it could prevent the imu from starting up. so it would be better to seperate those in their own files, and then have sensors launch file to launch all those launch files. very similar to the auv-2025 architecture.

JPGC04 commented on ros2_ws/src/sensors/params/sensors_frames.yaml:
this is very nice and it looks like an amazing way to modify params out of launch files

JPGC04 reviewed (dismissed):
this pr is great and should be merged. I unfortunately don't have enough technical knowledge to comment on the specifics of the code. I would hope this is tested in sim before it gets merged, but ultimately this branch needs to be merged when possible.

nadimasmar committed:
Added controls config files, preliminary tuning of attittude controller

nadimasmar committed:
Modified pid and depth controller logic.

ParseDotEXE commented on ros2_ws/src/controls/package.xml:
```diff
@@ -0,0 +1,26 @@
+  <description>Control nodes for the AUV.</description>
+
+  <maintainer email="dev@mcgillrobotics.com">McGill Robotics</maintainer>
+  <license>GPLv3</license>
```
good

ParseDotEXE commented on ros2_ws/src/controls/README.md:
```diff
@@ -0,0 +1,121 @@
+
+
+## Overview
+The controls package implements separate depth and attitude controllers. The depth controller uses a PID loop on `/processed/depth` and publishes a vertical effort on `/controls/depth_effort`. The attitude controller uses IMU orientation (`processed/imu`) and a target quaternion (`quaternion_setpoint`) to publish torques on `/controls/attitude_effort`. The superimposer node sums both efforts, app ...
```
great description

ParseDotEXE reviewed (commented):
done reviewing the PR! lmk if I missed anything or if you need anything to be cleared up

nadimasmar committed:
Tuned Depth controller in sim.

Stuart-Klenner committed:
Added comments

Stuart-Klenner committed:
Merge branch 'dev/llmweapons' into dev/dvl-integration

Stuart-Klenner committed:
Updated variable naming

nadimasmar committed:
Tuned attitude controller in sim

Alb33rt committed:
Merge branch 'main' into dev/controls

vbigattini committed:
Created planar controller branch

nadimasmar committed:
Added set param service to toggle controllers on and off

nadimasmar committed:
Update README

nadimasmar committed:
Update README

nadimasmar committed:
Merge branch 'dev/llmweapons' into dev/controls

nadimasmar committed:
Added depth calibration service.

nadimasmar committed:
Added depth calibration service.

nadimasmar committed:
Added target dependancy for state aggregator

nadimasmar committed:
Added target dependancy for state aggregator

nadimasmar committed:
Merge branch 'dev/llmweapons' into dev/controls

Alb33rt committed:
Merge branch 'main' into dev/controls

Stuart-Klenner committed:
Fixes for pull request

nadimasmar committed:
Merge branch 'dev/controls' into dev/planarcontroller

nadimasmar committed:
Added planar controller

Stuart-Klenner committed:
Removed nested class and decomposed ros msg's

Stuart-Klenner committed:
Update dvl_processor.cpp

Stuart-Klenner committed:
Update dvl_processor.cpp

nadimasmar committed:
add planar controller effort to superimposer and updated README

nadimasmar committed:
Merge branch 'dev/planarcontroller' into dev/controls

Stuart-Klenner committed:
added centralized frame id's

Stuart-Klenner committed:
Fixed dvl inertial frame data issues

Stuart-Klenner committed:
updating package dependancies

nadimasmar committed:
Merge branch 'dev/dvl-integration' into dev/llmweapons

nadimasmar committed:
Fixed CMake after faulty merge

nadimasmar committed:
Merge branch 'dev/llmweapons' into dev/controls

nadimasmar committed:
refactored DVL processing

nadimasmar committed:
corrected quaternion rotation from dvl to inertial and vehicle frames.

nadimasmar committed:
Added pool test tuning to config file. Added first guesses for other controllers

nadimasmar committed:
Update controller parameters and refine message handling in planar controller

- Increased KP value for x and y controllers from 10.0 to 15.0 for improved responsiveness.
- Updated subscription message type in planar controller to use PointStamped.
- Removed unnecessary logging statements in attitude and DVL processors to clean up output.

nadimasmar committed:
Updated documentation

nadimasmar committed:
Ensure error quaternions encodes shortest path
Co-authored-by: William Zhang <williamzhang2205@gmail.com>

nadimasmar committed:
Merge branch 'main' into dev/controls

nadimasmar committed:
DVL velocity transformation

nadimasmar committed:
Merge branch 'main' into dev/controls

nadimasmar committed:
Post-pool test improvments: Amended controller tuning, increased attitude controller frequency, changde param object tyoe to `double`, and updated documentation.

nadimasmar committed:
Disabled buoyancy torque correction.

WilliamZhang2205 committed:
Merge branch 'main' into dev/controls

WilliamZhang2205 reviewed (changes requested):
Overall very nice stuff! In the future, we can add doc strings above each function to facilitate the code reader's understandability. While, it is mentioned in the code to refer to the README.md, nice docstrings can be shown on your screen when hovering over a function making comprehension easier and more accessible. That can be done in another PR later on. The only mandatory change I'd ask is the removing the .vscode folder as JP mentioned before.

nadimasmar commented on ros2_ws/src/controls/src/superimposer.cpp:
```diff
@@ -0,0 +1,148 @@
+
+   void superimposer::attitude_effort_callback(const wrench_msg::SharedPtr msg)
+   {
+       attitude_effort_ = *msg; // This already is in the AUV body frame
```
This is a good idea. I will add this as an issue. New mech members will work on it. For this PR, I think we can let through the current version of superimposer.

However, what do you exactly mean by "reset". Do you mean to simply set the efforts to zero?

WilliamZhang2205 replied:
I mean we could have a service call equivalent to each effort callback. One service to set depth effort to 0, another to set x effort to 0, another for y effort to 0 and one last for attitude effort to 0.

nadimasmar replied:
I think this can already be achieved by setting the `enabled` parameter to `false` for every respective controller.

Only difference is that its a `ros2 param set` instead of a `ros2 service`.

WilliamZhang2205 replied:
I still believe it should be a service or a feature to set it to 0. At least maybe for X and Y controllers. For example, say we had set the setpoint to x=5, y =2, Dougie gets there, and we turn the controller off. If we turn it on again after having moved Dougie to x=0, y=0,  Dougie could start moving immediately and hurt someone. A fix could be when you set the param to false, the code could set those respective efforts to 0.

nadimasmar replied:
Say there is such a feature. When that feature is enabled, it overrides the controller logic and sets all efforts to zero. In the same scenario that you outline, disabling the feature would also cause the robot to jump suddenly, potentially hurting someone. 

if the `enabled` parameter is set to `false.` All controller efforts are set to zero and then published. Isn't that the same thing as what you are proposing?

nadimasmar committed:
Added VS code sttings to .gitignore

nadimasmar committed:
Remove .vscode from tracking

nadimasmar committed:
Limit parallel build level to 1 for debug builds to reduce memory usage. Track .vscode files.

avyan-k committed:
Adjust buoyancy parameters in sim config

avyan-k committed:
Merge branch 'main' into dev/controls

kaidalisohaib committed:
Add limit to planar effort to 20N (adjustable)

kaidalisohaib committed:
Merge branch 'dev/limit-planar-effort' into dev/controls

kaidalisohaib commented on ros2_ws/src/sensors/src/depth_processor.cpp:
```diff
@@ -0,0 +1,102 @@
+    r_vs_v_ = Vec3(r_vs_v_vec[0], r_vs_v_vec[1], r_vs_v_vec[2]);
+
+    depth_pub_ = this->create_publisher<float64_msg>("auv_frame/depth", 10);
+    depth_sub_ = this->create_subscription<float64_msg>(
+            "/sensors/depth/z",
+            10,
+            std::bind(&DepthProcessor::depth_callback, this, std::placeholders::_1)
+        );
+
+    imu_sub_ = this->create_subscription<sensor_msgs::msg::Imu>(
+        "auv_frame/imu",
+        10,
+        std::bind(&DepthProcessor::imu_callback, this, std::placeholders::_1)
+    );
```
For sensors subscribers, you should use rclcpp::SensorDataQoS() instead of 10 queue size to have the communication in "Best Effort" mode. 
Also, since the depth sensor is giving data at about 10 Hz (every 100ms), which is relatively slow, you would want to keep the queue size pretty small to keep the newest data. (rclcpp::SensorDataQoS().keep_last(2))
For imu, since it's faster it's ok to keep a queue size of 10, but you still want the best effort from the qos (rclcpp::SensorDataQoS().keep_last(10)).

nadimasmar replied:
Thank you for the suggestion. I implemented it. I set the queue size to be 1 for all topics. Once there is a newer reading, there is no value in getting the older one anymore.

kaidalisohaib commented on ros2_ws/src/sensors/src/dvl_processor.cpp:
```diff
@@ -0,0 +1,124 @@
+        "auv_frame/imu", 10, std::bind(&DvlProcessor::imu_callback, this, std::placeholders::_1));
+
+    dvl_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
+        "dvl/odometry", 10, std::bind(&DvlProcessor::dvl_callback, this, std::placeholders::_1));
```
Same idea for QoS of subscribers.
For imu rclcpp::SensorDataQoS().keep_last(10) is fine since it's very fast.
For dvl, 10 queue size at 5Hz (200 ms) is too much because it represents a potential delay of 2 seconds and since it's our first source of position we would want the latest data received. Therefore a queue size of 1 (rclcpp::SensorDataQoS().keep_last(1)).

nadimasmar replied:
Implemented with queue size 1

kaidalisohaib commented on ros2_ws/src/sensors/src/Imu_processor.cpp:
```diff
@@ -0,0 +1,123 @@
+        "auv_frame/imu",
+        10
+    );
+    imu_sub_ = this->create_subscription<imu_msg>(
+        "imu/data",
+        10,
+        std::bind(&ImuProcessor::imu_callback, this, std::placeholders::_1)
+    );
```
QoS update here as well.

nadimasmar replied:
Implemented with queue size 1

kaidalisohaib commented on ros2_ws/src/controls/scripts/depth_controller.py:
```diff
@@ -0,0 +1,109 @@
+        super().__init__('depth_controller')
+
+        # QoS: Always use most recent readings, reliable, volatile durability
+        qos = QoSProfile(
+            reliability=ReliabilityPolicy.RELIABLE,
+            durability=DurabilityPolicy.VOLATILE,
+            history=HistoryPolicy.KEEP_LAST,
+            depth=1,
+        )
+
+        self.pub_effort = self.create_publisher(Wrench, '/controls/depth_effort', qos)
+        self.sub_depth = self.create_subscription(Float64, 'auv_frame/depth', self.depth_callback, qos)
+        self.setpoint_sub = self.create_subscription(Float64, '/controls/depth_setpoint', self.setpoint_callback, qos)
```
Same idea as the other QoS update comments. For the sub_depth and pub_effort, you should switch from Reliable to Best Effort (using `qos_profile_sensor_data`). In a controller, it is better to drop a lost packet and move to the next one than wait for a late, old update. 
For the setpoints it should stay reliable since we want to make sure to receive the goals sent compared to sensors data which happens many times per seconds.

nadimasmar replied:
Done

kaidalisohaib commented on ros2_ws/src/controls/scripts/planar_controller.py:
```diff
@@ -0,0 +1,117 @@
+        
+        self.axis_name = axis_name
+
+        # QoS: Always use most recent readings, reliable, volatile durability
+        qos = QoSProfile(
+            reliability=ReliabilityPolicy.RELIABLE,
+            durability=DurabilityPolicy.VOLATILE,
+            history=HistoryPolicy.KEEP_LAST,
+            depth=1,
+        )
+
+        self.pub_effort = self.create_publisher(Wrench, f'/controls/{axis_name}_effort', qos)
+        self.sub_position = self.create_subscription(PointStamped, f'auv_frame/dvl/position', self.position_callback, qos)
+        self.setpoint_sub = self.create_subscription(Float64, f'/controls/{axis_name}_setpoint', self.setpoint_callback, qos)
```
Same idea as the other QoS comments. Except for setpoint which should stay at reliable.

nadimasmar replied:
Done

kaidalisohaib commented on ros2_ws/src/controls/src/attitude_controller.cpp:
```diff
@@ -0,0 +1,188 @@
+
+
+        pub_effort_ = this->create_publisher<wrench_msg>("/controls/attitude_effort", 1);
+        sub_imu_ = this->create_subscription<imu_msg>(
+            "auv_frame/imu",
+            1,
+            std::bind(&AttitudeController::imu_callback, this, std::placeholders::_1)
+        );
+        sub_target_orientation_ = this->create_subscription<geometry_msgs::msg::Quaternion>(
+            "/controls/quaternion_setpoint",
+            1,
+            std::bind(&AttitudeController::target_orientation_callback, this, std::placeholders::_1)
+        );
```
For the IMU and effort topics, we should use rclcpp::SensorDataQoS() instead of a queue size of 1. Even though it seems counter-intuitive, Reliable mode (the default for 1) is actually slower because it pauses to retry lost packets. In a high-speed attitude loop, an old IMU reading is useless. It’s better to drop a lost packet and move immediately to the next one to keep the controller stable.

nadimasmar replied:
Done

kaidalisohaib commented on ros2_ws/src/controls/src/attitude_controller.cpp:
```diff
@@ -0,0 +1,188 @@
+                0, 0, P_wz_;
+
+
+        pub_effort_ = this->create_publisher<wrench_msg>("/controls/attitude_effort", 1);
```
For pub_effort, it's better to use Best Effort (rclcpp::SensorDataQoS()). If an effort topic is lost, we don't want the system to waste time retrying it. By the time the 'retry' arrives, it's already out of date. It's much safer for the thrusters to just receive the next fresh calculation immediately to keep the control loop smooth.

nadimasmar replied:
Done

kaidalisohaib commented on ros2_ws/src/controls/src/superimposer.cpp:
```diff
@@ -0,0 +1,160 @@
+       );
+
+       // Initialize publisher
+       pub_effort_ = this->create_publisher<wrench_msg>("/controls/total_effort", 1);
```
Same idea for QoS as the other comments.

nadimasmar replied:
Done

kaidalisohaib commented on ros2_ws/src/sensors/src/State_aggregator.cpp:
```diff
@@ -0,0 +1,122 @@
+
+
+
+        imu_sub_ = this->create_subscription<imu_msg>(
+            "auv_frame/imu",
+            1, // Use queue size 1 to get latest data
+            std::bind(&State_aggregator::imu_callback, this, std::placeholders::_1)
+        );
+        depth_sub_ = this->create_subscription<float64_msg>(
+            "auv_frame/depth",
+            1,
+            std::bind(&State_aggregator::depth_callback, this, std::placeholders::_1)
+        );
+
+        dvl_position_sub_ = this->create_subscription<geometry_msgs::msg::PointStamped>(
+            "auv_frame/dvl/position",
+            1,
+            std::bind(&State_aggregator::dvl_position_callback, this, std::placeholders::_1)
+        );
+
+        dvl_velocity_sub_ = this->create_subscription<geometry_msgs::msg::TwistStamped>(
+            "auv_frame/dvl/velocity",
+            1,
+            std::bind(&State_aggregator::dvl_velocity_callback, this, std::placeholders::_1)
+        );
+
+        pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>(
+            "state/pose",
+            1
+        );
```
Same QoS updates here as well.

nadimasmar replied:
Done

kaidalisohaib reviewed (changes requested):
For sensor subscribers, we should switch from Reliable to Best Effort using `rclcpp::SensorDataQoS()` or `from rclpy.qos import qos_profile_sensor_data` in python or a custom QoSProfile.
This profile is made specifically for sensors, it stops the system from retrying lost packets, which avoids lag. For our use case getting the newest data quickly is more important than making sure every single old packet arrives.

kaidalisohaib committed:
Merge branch 'main' into dev/controls

nadimasmar committed:
Update QoS profiles

avyan-k committed:
Update git submodule URL for Xsens ROS Driver

kaidalisohaib committed:
Fix incompatible sub/pub reliability (best effort pub and reliable sub)

kaidalisohaib reviewed (approved):
Tested in the sim 👍 
Note: Hopefully simply tuning the controllers once again will fix the instability seen during pre-qual.
